### PR TITLE
Migration to Rice 4

### PR DESCRIPTION
**Pull request by alemorozov**
This is an upgrade of the generator to the rice 4.0 interface.

It uses the latest rice and, by the way, properly detects the `arch` attribute https://github.com/metanorma/packed-mn/issues/140.

### CODE REVIEW

File: lib/antlr4-native/generator.rb [hunk 1]
```diff
@@ -65,47 +65,60 @@ def interop_code
...
 
         #ifdef _WIN32
-        #undef FALSE
-        #undef TRUE
         #undef OPTIONAL
         #undef IN
         #undef OUT
         #endif
 
+        // ruby-3.0
```
Comment (camertron) [inline lib/antlr4-native/generator.rb]: Hmm so does this update work for older Ruby versions then?

**Review by camertron — COMMENTED** [review]
I'm concerned these changes will break older Ruby versions, i.e. with the change from `define_singleton_method` to `define_singleton_function`, etc. Have you tested any Ruby 2 versions?

**Comment by alemorozov** [comment]
It works with all Rubies starting from 2.5. We tested it on the [expressir](https://github.com/lutaml/expressir) library [here](https://github.com/lutaml/expressir/runs/4851472857?check_suite_focus=true). A generated by antlr4-native code is [here](https://github.com/lutaml/expressir/blob/main/ext/express-parser/express_parser.cpp) (warning, it's a pretty large file).

Also added a dependency on a new version (4.0) of rice, so a new antlr4-native is used only when a project uses a new rice.

**Comment by camertron** [comment]
Ok awesome! Do you think I should release a major version bump then? Also, if you have any ideas around reducing generated code size, I'd love to hear them.

**Comment by alemorozov** [comment]
Since output changes are quite big, I think, it can be a major version bump. So the previous major release can be kept for Rice 3.0 and its bug fixes.

Ok :)

Thank you @camertron 🙏

**Comment by camertron** [comment]
Hey @alexeymorozov I tried re-generating the code for [python3-parser-rb](https://github.com/camertron/python3-parser-rb) using your changes. Everything compiles just fine, but the tests fail with "Bad cast" exceptions. Do you have any idea what might be going on? I haven't been able to pinpoint the cause, but it appears to happen in the visitor code somewhere. You can reproduce by running `rake generate`, then `rake compile`, then `bundle exec rspec`.

**Comment by alemorozov** [comment]
Still haven't dig much further, but it fails the same way with [Rice 2.0](https://github.com/alexeymorozov/python3-parser-rb/pull/1/checks).

To compare with [Rice 4.0](https://github.com/alexeymorozov/python3-parser-rb/pull/2/checks) build.

**Comment by camertron** [comment]
Hey @alexeymorozov, I was able to dig in and I'm pretty sure I know what happened here. Several PRs from outside contributors were merged a while ago and I never regenerated the C++ code for python3-parser. One of the PRs introduced the `std::bad_cast` bug, which I believe is entirely unrelated to the upgrade to Rice v4.0. The issue should be fixed in master. Feel free to merge or rebase whenever you have the chance :)

**Comment by alemorozov** [comment]
I'm glad it's fixed now! Rebased the branch. Tests [pass](https://github.com/alexeymorozov/python3-parser-rb/pull/2/checks) with python3-parser.

**Comment by camertron** [comment]
Released in v2.0.0 :)

**Comment by alemorozov** [comment]
Thank you @camertron 😄

**Pull request merged**